vllm/v1/core/scheduler.py

```diff
@@ -480,23 +518,47 @@ def update_from_output(
             scheduler_stats=self.make_stats(),
         )
 
-    def _check_stop(self, request: Request) -> bool:
+    def _maybe_stop_and_crop(self, request: Request) -> bool:
```

Comment:
This method looks inefficient because it searches in O(n) time. Can we only consider the most recently generated tokens?

Reply:
Now that these checks are not quite as trivial, we could move them to the front-end process like we do with stop strings (at least the stop token checks).

Reply:
The reason behinds this is now draft model won't stop even stop token is generated so the stop token can be in any position. The fix is supposed to be a follow-up PR, but if this is a blocker we should add it in this PR. cc @LiuXiaoxuanPKU

Reply:
@comaniac no I think @WoosukKwon is saying that we only need to scan the newly generated/speculated tokens here rather than the entire prompt.

I actually think it's probably better not to have the proposer care about stop tokens.

Reply:
Why proposer should not consider stop token? In this case if the first proposed token is a stop token, then the rest proposed tokens are wasted for sure.

Reply:
I just thought it might be simpler to have the checks in one place and it wouldn't make much difference to overall performance if one or two additional tokens are proposed at the end. But I guess it wouldn't really add much complexity.

Reply:
@LiuXiaoxuanPKU and I discussed this offline. We found that the logic is actually simple: we just need a python loop iterating over each sampled token id (instead of adding all token ids into `output_token_ids` and then checking the stop condition):

```python
for token_id in sampled_token_ids:
    stopped = self.check_stop(request, token_id)
    request.append_output_token_ids(token_id)
    ...
    if stopped:
        break
```